crates/bevy_ecs/src/storage/resource.rs

```diff
@@ -2,19 +2,55 @@ use crate::archetype::ArchetypeComponentId;
 use crate::component::{ComponentId, ComponentTicks, Components, TickCells};
 use crate::storage::{Column, SparseSet, TableRow};
 use bevy_ptr::{OwningPtr, Ptr, UnsafeCellDeref};
+use std::{mem::ManuallyDrop, thread::ThreadId};
 
 /// The type-erased backing storage and metadata for a single resource within a [`World`].
 ///
+/// If `SEND` is false, values of this type will panic if dropped from a different thread.
+///
 /// [`World`]: crate::world::World
-pub struct ResourceData {
-    column: Column,
+pub struct ResourceData<const SEND: bool> {
+    column: ManuallyDrop<Column>,
+    type_name: String,
     id: ArchetypeComponentId,
+    origin_thread_id: Option<ThreadId>,
+}
+
+impl<const SEND: bool> Drop for ResourceData<SEND> {
+    fn drop(&mut self) {
+        if self.is_present() {
+            self.validate_access();
+        }
+        // SAFETY: Drop is only called once upon dropping the ResourceData
+        // and is inaccessible after this as the parent ResourceData has
+        // been dropped.
+        unsafe {
+            ManuallyDrop::drop(&mut self.column);
+        }
+    }
 }
 
-impl ResourceData {
+impl<const SEND: bool> ResourceData<SEND> {
     /// The only row in the underlying column.
     const ROW: TableRow = TableRow::new(0);
 
+    #[inline]
+    fn validate_access(&self) {
+        // Avoid aborting due to double panicking on the same thread.
+        if SEND || std::thread::panicking() {
```

Comment:
We can't test for this, at least not without explicitly shielding the test via a subprocess. Panicking while panicking will abort.

Reply:
We could switch this back to the original approach and just hard-abort instead of panicking. It'd be equally as hard to test, but there's zero ambiguity or corner cases to check for here.

Reply:
Did this last option.